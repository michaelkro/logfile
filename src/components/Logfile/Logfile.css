#root {
  padding: 2rem;
  height: 100vh;
}

.logfile-header {
  display: flex;
  align-items: center;
}

.logfile-header__brand {
  margin-right: var(--spacing-xs);
}

main {
  display: flex;
  flex-direction: column;
}

.logfile-url-form {
  align-items: flex-start;
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-bottom: var(--spacing-lg);
}

.logfile-url-form__input {
  display: flex;
  width: 100%;
}

.logfile-url-form__input input {
  flex: auto;
  border: 1px solid #ddd;
  border-radius: var(--border-radius);
  margin-right: var(--spacing-sm);
  padding: 0 var(--spacing-xs);
  box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.075);
  line-height: 2;
  font-size: 1rem;
}

.logfile-url-form__input button {
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: var(--border-radius);
  width: 90px;
  padding: var(--spacing-xs) var(--spacing-sm);
  background-color: #0fb9b8;
  color: white;
  cursor: pointer;
  font-size: 1rem;
}

.logfile-url-form__input button:disabled {
  background-color: #9fd3d3;
  color: #e0e0e0;
  cursor: not-allowed;
  opacity: 0.7;
  border: 1px solid #cccccc;
  pointer-events: none;
}

.logfile-url-form__input__error {
  padding: 0 var(--spacing-xs);
  color: red;
  font-size: 0.9rem;
}

.logfile-table {
  /* header - input - input margin - table header - root padding */
  height: calc(100vh - 90px - 55px - 55px - 40px - 30px);
}

.logfile-table__header {
  display: flex;
  background-color: #f0f0f0;
  font-weight: bold;
  border-top-left-radius: var(--border-radius);
  border-top-right-radius: var(--border-radius);
}

.logfile-table__header > div {
  padding: var(--spacing-xs) var(--spacing-sm)
}

.logfile-table__header__time-cell {
  width: 225px;
}

.logfile-table__body {
  overflow-y: scroll;
  height: 100%;
  margin: 0;
  padding: 0;
  border: 1px solid #ddd;
  white-space: nowrap;
  border-bottom-right-radius: var(--border-radius);
  border-bottom-left-radius: var(--border-radius);
}
