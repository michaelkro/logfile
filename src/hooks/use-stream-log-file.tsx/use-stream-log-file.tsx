import { useCallback, useRef, useState } from 'react'

import { Logger } from '../../services/logger'

import { NEWLINE_REGEX, URL_REGEX, MAX_FILE_SIZE } from './constants'

interface UseStreamLogFileOptions {
  maxFileSize?: number
}

export const useStreamLogFile = (options?: UseStreamLogFileOptions) => {
  const [error, setError] = useState('')
  const [loading, setLoading] = useState(false)
  const [logLines, setLoglines] = useState<string[]>([])
  const lineBufferRef = useRef('')
  const totalSizeRef = useRef<number>(0)
  const maxFileSize = options?.maxFileSize || MAX_FILE_SIZE

  const clearError = useCallback(() => {
    setError('')
  }, [])

  const streamLogfile = useCallback(
    async (url: string) => {
      try {
        if (!url || !URL_REGEX.test(url)) {
          throw new Error('Must provide a valid web URL')
        }

        setLoading(true)
        setError('')
        setLoglines([])
        lineBufferRef.current = ''
        totalSizeRef.current = 0

        const response = await fetch(url, {
          headers: { Accept: 'application/json' }
        })

        if (!response.ok) {
          throw new Error(`${response.status} ${response.statusText}`)
        }

        if (!response.body) {
          throw new Error('Response body is null')
        }

        const reader = response.body
          .pipeThrough(new TextDecoderStream())
          .getReader()

        while (true) {
          const { done, value: chunk } = await reader.read()

          if (done) {
            if (lineBufferRef.current) {
              // Handle any remaining text in the buffer
              setLoglines((logLines) => [...logLines, lineBufferRef.current])
              lineBufferRef.current = ''
            }
            break
          }

          // If file is too big, stop processing and capture an event
          totalSizeRef.current = totalSizeRef.current + chunk.length
          if (totalSizeRef.current > maxFileSize) {
            Logger.info('Max file size exceeded', {
              maxFileSize: maxFileSize.toString(),
              fileSize: totalSizeRef.current.toString()
            })
            setError(
              `The file size is greater than the allowed maximum of ${maxFileSize}. The full log file cannot be displayed.`
            )
            setLoading(false)
            break
          }

          // Add back leftover text from previous chunk
          const text = lineBufferRef.current + chunk
          const lines = text.split(NEWLINE_REGEX)

          // Could be an incomplete line
          lineBufferRef.current = lines.pop() || ''

          if (lines.length > 0) {
            setLoglines((prevState) => [...prevState, ...lines])
          }
        }
      } catch (err) {
        if (err instanceof Error) {
          setError(`Failed to fetch log file: ${err.message}`)
        } else {
          throw err
        }
      } finally {
        setLoading(false)
      }
    },
    [maxFileSize]
  )

  return {
    clearError,
    error,
    loading,
    logLines,
    streamLogfile
  } as const
}
