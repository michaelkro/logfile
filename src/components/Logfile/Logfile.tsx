import React, { useState, useEffect } from 'react'

import { useStreamLogFile } from '../../hooks/use-stream-log-file.tsx'

import LogfileIcon from '../../../public/logfile.svg'

import './Logfile.css'

interface LogData {
  [key: string]: string
  _time: string
}

interface LogEntry {
  timestamp?: string
  raw: string
  error?: string
}

export function Logfile() {
  const [url, setUrl] = useState<string>('')
  const [logs, setLogs] = useState<LogEntry[]>([])
  const { streamLogfile, logLines, error } = useStreamLogFile()

  useEffect(() => {
    if (logLines && logLines.length > 0) {
      const parsedLogs: LogEntry[] = logLines.map((logLine) => {
        try {
          const json: LogData = JSON.parse(logLine)
          return { timestamp: new Date(json._time).toISOString(), raw: logLine }
        } catch {
          return { raw: logLine, error: 'Invalid JSON' }
        }
      })

      setLogs(parsedLogs)
    }
  }, [logLines])

  return (
    <>
      <header className="logfile-header">
        <img
          className="logfile-header__brand"
          src={LogfileIcon}
          alt="logfile-svg"
        />
        <h1>logfile</h1>
      </header>
      <main>
        <form
          className="logfile-url-form"
          onSubmit={(e) => {
            e.preventDefault()
            streamLogfile(url)
          }}
        >
          <div className="logfile-url-form__input">
            <input
              placeholder="Enter download URL for your log file"
              value={url}
              onChange={(e) => setUrl(e.target.value)}
            />
            <button type="submit">Submit</button>
          </div>
          {error && (
            <span className="logfile-url-form__input__error">{error}</span>
          )}
        </form>
        {logs.length > 0 && (
          <div className="logfile-table">
            <div className="logfile-table__header">
              <div className="logfile-table__header__time-cell">Time</div>
              <div className="logfile-table__header__event-cell">Event</div>
            </div>
            <div className="logfile-table__body">
              {logs.map((log, index) => {
                return (
                  <div className="logfile-table__row" key={index}>
                    <div className="logfile-table__row__time-cell">
                      {log.timestamp}
                    </div>
                    <div className="logfile-table__row__event-cell">
                      {log.raw}
                    </div>
                  </div>
                )
              })}
            </div>
          </div>
        )}
      </main>
    </>
  )
}
