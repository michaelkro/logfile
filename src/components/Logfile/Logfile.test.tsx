import { render, screen, fireEvent } from '@testing-library/react'

import { useStreamLogFile } from '../../hooks/use-stream-log-file.tsx'

import { Logfile } from './Logfile'

const LOGS = [
  '{"_time": "2023-04-01T12:00:00Z", "level": "info", "message": "Log message 1"}',
  '{"_time": "2023-04-01T12:01:00Z", "level": "error", "message": "Log message 2"}'
]

jest.mock('../../hooks/use-stream-log-file.tsx', () => ({
  useStreamLogFile: jest.fn()
}))

test('renders logs that are returned', () => {
  ;(useStreamLogFile as jest.Mock).mockReturnValue({
    logLines: LOGS,
    clearError: jest.fn()
  })

  render(<Logfile />)

  const table = screen.getByRole('table')

  expect(table).toMatchSnapshot()
})

test('clicking a row shows expanded view with formatted JSON', () => {
  const LOG = LOGS[0]
  ;(useStreamLogFile as jest.Mock).mockReturnValue({
    logLines: [LOG],
    clearError: jest.fn()
  })

  render(<Logfile />)

  const row = screen.getByRole('row')
  fireEvent.click(row)

  expect(row).toMatchSnapshot()
})

test('submit button is disabled when input is empty', () => {
  ;(useStreamLogFile as jest.Mock).mockReturnValue({
    clearError: jest.fn()
  })
  render(<Logfile />)
  const submitButton = screen.getByRole('button', { name: /submit/i })

  expect(submitButton).toBeDisabled()
})

test('shows an error message if an error is returned ', () => {
  const errorMessage = 'Whoops'
  ;(useStreamLogFile as jest.Mock).mockReturnValue({
    clearError: jest.fn(),
    error: errorMessage
  })

  render(<Logfile />)

  expect(screen.getByText(errorMessage)).toBeInTheDocument()
})

test('button shows spinner when logs are loading', () => {
  ;(useStreamLogFile as jest.Mock).mockReturnValue({
    loading: true,
    clearError: jest.fn()
  })

  render(<Logfile />)

  expect(screen.getByTestId('loading-spinner')).toBeInTheDocument()
})
