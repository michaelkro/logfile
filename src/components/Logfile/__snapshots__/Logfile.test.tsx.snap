// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking a row shows expanded view with formatted JSON 1`] = `
<div
  class="logfile-table__row expanded"
  role="row"
  style="transform: translateY(0px); height: 80px;"
>
  <div>
    <div
      class="logfile-table__row__time-cell expanded"
    >
      2023-04-01T12:00:00.000Z
    </div>
    <pre
      class="logfile-table__row__formatted-json"
    >
      {
  "_time": "2023-04-01T12:00:00Z",
  "level": "info",
  "message": "Log message 1"
}
    </pre>
  </div>
</div>
`;

exports[`renders logs that are returned 1`] = `
<div
  class="logfile-table"
  role="table"
>
  <div
    class="logfile-table__header"
  >
    <div
      class="logfile-table__header__time-cell"
    >
      Time
    </div>
    <div
      class="logfile-table__header__event-cell"
    >
      Event
    </div>
  </div>
  <div
    class="logfile-table__body"
  >
    <div
      style="height: 80px; position: relative; width: 5000px;"
    >
      <div
        class="logfile-table__row"
        role="row"
        style="transform: translateY(0px); height: 40px;"
      >
         
        <div
          class="logfile-table__row__time-cell"
        >
          2023-04-01T12:00:00.000Z
        </div>
        <div
          class="logfile-table__row__event-cell"
        >
          {"_time": "2023-04-01T12:00:00Z", "level": "info", "message": "Log message 1"}
        </div>
      </div>
      <div
        class="logfile-table__row"
        role="row"
        style="transform: translateY(40px); height: 40px;"
      >
         
        <div
          class="logfile-table__row__time-cell"
        >
          2023-04-01T12:01:00.000Z
        </div>
        <div
          class="logfile-table__row__event-cell"
        >
          {"_time": "2023-04-01T12:01:00Z", "level": "error", "message": "Log message 2"}
        </div>
      </div>
    </div>
  </div>
</div>
`;
